lang/py/src/avro/io.py

```diff
@@ -407,53 +394,60 @@ def write_decimal_bytes(self, datum, scale):
     Decimal in bytes are encoded as long. Since size of packed value in bytes for
     signed long is 8, 8 bytes are written.
     """
-    if isinstance(datum, Decimal):
-      datum = datum.to_eng_string()
-    else:
-      datum = str(datum)
-    decimal_place = datum.find('.')
-    if decimal_place == -1:
-      mantissa = 0
-    else:
-      mantissa = len(datum[decimal_place+1:])
-    if mantissa != scale:
+    sign, digits, exp = datum.as_tuple()
+    if exp > scale:
       raise AvroTypeException('Scale provided in schema does not match the decimal')
-    unscaled_datum = datum.replace('.', '')
-    unscaled_int_datum = long(unscaled_datum)
-    packed_bytes = STRUCT_SIGNED_LONG.pack(unscaled_int_datum)
-    self.write_long(8)
-    for i in range(8):
-      self.write(packed_bytes[i])
+
+    unscaled_datum = 0
+    for digit in digits:
+      unscaled_datum *= 10
+      unscaled_datum += digit
+
+    bits_req = unscaled_datum.bit_length()
+    if sign:
+      bits_req += 1
+      unscaled_datum = (1 << bits_req) - unscaled_datum
+
+    bytes_req = bits_req // 8
+    padding_bits = ~((1 << bits_req) - 1) if sign else 0
+    packed_bits = padding_bits | unscaled_datum
+
+    bytes_req += 1 if (bytes_req << 3) < bits_req else 0
+    self.write_long(bytes_req)
+    for index in range(bytes_req-1, -1, -1):
+      bits_to_write = packed_bits >> (8 * index)
+      self.write(chr(bits_to_write & 0xff))
 
   def write_decimal_fixed(self, datum, scale, size):
     """
     Decimal in fixed are encoded as size of fixed bytes.
     """
-    if isinstance(datum, Decimal):
-      datum = datum.to_eng_string()
-    else:
-      datum = str(datum)
-    decimal_place = datum.find('.')
-    if decimal_place == -1:
-      mantissa = 0
-    else:
-      mantissa = len(datum[decimal_place+1:])
-    if mantissa != scale:
+    sign, digits, exp = datum.as_tuple()
+    if exp > scale:
       raise AvroTypeException('Scale provided in schema does not match the decimal')
-    unscaled_datum = datum.replace('.', '')
-    unscaled_long_datum = long(unscaled_datum)
-    if unscaled_long_datum < 0:
-      fill_byte = '\xff'
-    else:
-      fill_byte = '\x00'
-    packed_bytes = STRUCT_SIGNED_LONG.pack(unscaled_long_datum)
-    offset_bytes = size - len(packed_bytes)
-    padded_bytes = (fill_byte * offset_bytes) + packed_bytes
-    for i in range(size):
-      if padded_bytes[i] == '\x00':
-        self.write(chr(0))
-      else:
-        self.write(padded_bytes[i])
+
+    unscaled_datum = 0
+    for digit in digits:
+      unscaled_datum *= 10
+      unscaled_datum += digit
+
+    bits_req = unscaled_datum.bit_length()
+    if sign:
+      bits_req += 1
+      unscaled_datum = (1 << bits_req) - unscaled_datum
+
+    bytes_req = bits_req // 8
+    bytes_req += 1 if (bytes_req << 3) < bits_req else 0
```

Comment:
The bytes_req should be the given fixed size. It is possible the number of padding bits are more than 1 byte.